static/app/components/dropdownMenuV2.tsx

```diff
@@ -0,0 +1,301 @@
+  });
+
+  /**
+   * Close the menu on outside interaction, blur, or Esc key press,
+   * and control its position relative to the trigger button.
+   * See:
+   *   https://react-spectrum.adobe.com/react-aria/useOverlay.html
+   *   https://react-spectrum.adobe.com/react-aria/useOverlayPosition.html
+   */
+  const overlayRef = useRef(null);
+  const {overlayProps} = useOverlay(
+    {
+      onClose: closeRootMenu,
+      shouldCloseOnBlur,
+      isDismissable,
+      isOpen: true,
+    },
+    overlayRef
+  );
+  const {overlayProps: positionProps, placement: placementProp} = useOverlayPosition({
+    targetRef: triggerRef,
+    overlayRef,
+    offset,
+    crossOffset,
+    placement,
+    containerPadding,
+    isOpen: true,
+  });
+
+  /**
+   * Store whether this menu/submenu is the current focused one, which
+   * in a nested, tree-like menu system should be the leaf submenu.
+   * This information is used for controlling keyboard events.
+   * See: modifiedMenuProps below.
+   */
+  const [hasFocus, setHasFocus] = useState(true);
+  useEffect(() => {
+    /**
+     * A submenu is a leaf when it does not contain any expanded submenu.
+     * This logically follows from the tree-like structure and single-selection
+     * nature of menus.
+     */
+    const isLeafSubmenu = !stateCollection.some(node => {
+      const isSection = node.hasChildNodes && !node.value.isSubmenu;
+      /**
+       * A submenu with key [key] is expanded if
+       * state.selectionManager.isSelected([key]) = true
+       */
+      return isSection
+        ? [...node.childNodes].some(child =>
+            state.selectionManager.isSelected(`${child.key}`)
+          )
+        : state.selectionManager.isSelected(`${node.key}`);
+    });
+    setHasFocus(isLeafSubmenu);
+  }, [state.selectionManager.selectedKeys]);
+  /**
+   * Menu props from useMenu, modified to disable keyboard events
+   * if the current menu does not have focus.
+   */
+  const modifiedMenuProps = {
+    ...menuProps,
+    ...(!hasFocus && {
+      onKeyUp: () => null,
+      onKeyDown: () => null,
+    }),
+  };
+
+  /**
+   * Render a single menu item
+   */
+  const renderItem = (node: Node<MenuItemProps>, isLastNode: boolean) => {
+    return (
+      <MenuItem
+        node={node}
+        isLastNode={isLastNode}
+        state={state}
+        onClose={closeRootMenu}
+        closeOnSelect={closeOnSelect}
+      />
+    );
+  };
+
+  /**
+   * Render a submenu whose trigger button is a menu item
+   */
+  const renderItemWithSubmenu = (node: Node<MenuItemProps>, isLastNode: boolean) => {
+    const trigger = ({props: submenuTriggerProps, ref: submenuTriggerRef}) => (
+      <MenuItem
+        ref={submenuTriggerRef}
+        renderAs="div"
+        node={node}
+        isLastNode={isLastNode}
+        state={state}
+        isSubmenuTrigger
+        {...submenuTriggerProps}
+      />
+    );
+
+    return (
+      <MenuControl
+        items={node.value.children as MenuItemProps[]}
```

Comment:
Why does it need the case?

Reply:
without it, typescript wouldn't understand that this code is only reachable when `node.value.children` is defined and is an array.

<img width="815" alt="Screen Shot 2022-01-28 at 2 14 40 PM" src="https://user-images.githubusercontent.com/44172267/151628933-9b6d7a99-a887-49bf-8168-858ac60a3921.png">